docs/content/sips/007-deployment-auth.md

```diff
@@ -0,0 +1,168 @@
+title = "SIP 007 - Deployment Authentication and Authorisation"
+template = "main"
+date = "2022-09-30T01:01:01Z"
+---
+
+Summary: User experience and implementation design for authentication to deployment targets.
+
+Owners: ivan.towlson@fermyon.com
+
+Created: Sep 30, 2022
+
+## Background
+
+`spin deploy` currently requires the user to authenticate on every deployment.  For a "username-password" style of authentication, this is relatively painless as the user can set the `HIPPO_USERNAME` and `HIPPO_PASSWORD` environment variables and never need to worry about it.  However, it's unsatisfactory for token-based schemes such as OAuth or device code login, which are modelled more around a login experience that grants an access token.
+
+We want to provide a way for `spin deploy` to work smoothly against environments configured with these schemes, as well as the existing "username-password" environments.
+
+## Current experience
+
+To recap, the current `spin deploy` authentication experience is:
+
+* `--bindle-server` or `$BINDLE_URL` - required
+* `--bindle-username` or `$BINDLE_USERNAME` - optional
+* `--bindle-password` or `$BINDLE_PASSWORD` - optional
+* `--hippo-server` or `$HIPPO_URL` - required
+* `--hippo-username` or `$HIPPO_USERNAME` - required
+* `--hippo-password` or `$HIPPO_PASSWORD` - required
+
+## Proposal
+
+### Conceptual model
+
+At any given time, Spin is either be _logged out_ (no active deployment target, no credentials stored for any target) or _logged in_ to a deployment target (URL and credentials stored for that target).  Logging in is a separate step from deployment.  Deployment requires Spin to be logged in, and always deploys to the stored URL with the stored credentials
+
+**NOTE:** This will apply to existing username-password environments as well.  The current UI for deploying to those environments goes away.
+
+**FUTURE:** In future Spin may be able to store multiple deployment targets, in a similar manner to Kubernetes _contexts_.  A user would then be able to switch between these contexts without re-establishing credentials.  It would be desirable to have some level of this very early on, so that users or Fermyon staff could easily test on deployment targets running stable and canary Spin versions.
+
+### Login command
+
+We propose to add a `spin login` command, which takes the desired Hippo server as an argument:
+
+`spin login --url https://example.com`
+
+This launches an interactive login experience as follows.
+
+**1. Determine auth type.**
+
+At the moment, Spin doesn't know whether a URL uses device flow or username-password, so `spin login` will prompt.  If we can identify particular URLs or families of URLs whose auth mode is known, those can skip this step.
+
+```
+What authentication method does this server support?
+1. Sign in with GitHub
+2. Sign in with a username and password
+Enter a number:
+```
+
+Programmatic consumers need a way to override this prompt, e.g. `--method=[username|github]`.
```

Comment:
I think this would be a good option. We can also check for the presence of `--check-device-code` to determine the auth mode, since the presence of that flag automatically short-circuits it to be Github auth.

Reply:
Yep, and `--get-device-code` too of course.  I'll add this note and merge.  Thanks!